internal/alb/ls/listener.go

```diff
@@ -328,8 +333,113 @@ func (l *Listener) GetRules() rs.Rules {
 }
 
 func (l *Listener) DefaultActionArn() *string {
-	if *l.ls.current.DefaultActions[0].Type == elbv2.ActionTypeEnumRedirect {
-		return l.ls.current.DefaultActions[0].TargetGroupArn
+	if l.ls.current == nil {
+		l.logger.Debugf("DefaultAction: l.ls.current is nil")
+		return nil
+
+	} else if len(l.ls.current.DefaultActions) > 0 {
+		l.logger.Debugf("DefaultAction: l.ls.current.DefaultActions is empty")
+		return nil
+
+	} else if l.ls.current.DefaultActions[0] != nil {
+		l.logger.Debugf("DefaultAction: l.ls.current.DefaultActions[0] is nil")
+		return nil
+
+	} else if *l.ls.current.DefaultActions[0].Type != elbv2.ActionTypeEnumRedirect {
+		l.logger.Debugf("DefaultAction: l.ls.current.DefaultActions[0].Type is not 'redirect'")
+		return nil
 	}
-	return nil
+
+	return l.ls.current.DefaultActions[0].TargetGroupArn
+}
+
+func getCertificates(certificateArn *string, ingress *extensions.Ingress, logger *log.Logger) ([]*elbv2.Certificate, error) {
+	if certificateArn != nil {
+		logger.Debugf("New desired listener has certificate-arn '%v' in annotation", certificateArn)
+		return []*elbv2.Certificate{
+			{CertificateArn: certificateArn},
+		}, nil
+	}
+
+	logger.Debugf("New desired listener wants HTTPS, but hasn't provided an certificate-arn annotation")
+
+	var input = &acm.ListCertificatesInput{
+		CertificateStatuses: aws.StringSlice([]string{acm.CertificateStatusIssued}),
+
+		// AWS documentation doesn't specify what the actual default is
+		MaxItems: aws.Int64(500),
+	}
+	var certs []*elbv2.Certificate
+	var seen = map[string]bool{}
+	var page = 0
+	var ingressHosts = uniqueHosts(ingress)
+	err := albacm.ACMsvc.ListCertificatesPages(input, func(output *acm.ListCertificatesOutput, _ bool) bool {
+		logger.Debugf("%d issued certificates in AWS ACM response page %d", len(output.CertificateSummaryList), page)
+		for _, c := range output.CertificateSummaryList {
+			for _, h := range ingressHosts {
+				if domainMatchesIngressTLSHost(aws.StringValue(c.DomainName), h) {
+					logger.Debugf("Domain name '%s', matches TLS host '%v', adding to Listener", aws.StringValue(c.DomainName), h)
+					if !seen[aws.StringValue(c.CertificateArn)] {
+						certs = append(certs, &elbv2.Certificate{CertificateArn: c.CertificateArn})
+						seen[aws.StringValue(c.CertificateArn)] = true
+					}
+				} else {
+					logger.Debugf("Ignoring domain name '%s', doesn't match '%s'", aws.StringValue(c.DomainName), h)
+				}
+			}
+		}
+		page++
+		return true
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return certs, nil
+}
+
+func domainMatchesIngressTLSHost(domainName string, tlsHost string) bool {
```

Comment:
Should `TLS` be removed from this function name given that we're now using both TLS hosts and Rules hosts?

Reply:
Good call!